// elysia/dist/cjs/type-system.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.t = exports.ElysiaType = void 0;
const system_1 = require("@sinclair/typebox/system");
const typebox_1 = require("@sinclair/typebox");
const value_1 = require("@sinclair/typebox/value");
const error_1 = require("./error");
const t = Object.assign({}, typebox_1.Type);
exports.t = t;
try {
    system_1.TypeSystem.Format('email', (value) => /^[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i.test(value));
    system_1.TypeSystem.Format('uuid', (value) => /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value));
    system_1.TypeSystem.Format('date', (value) => !Number.isNaN(new Date(value).getTime()));
    system_1.TypeSystem.Format('date-time', (value) => !Number.isNaN(new Date(value).getTime()));
}
catch (error) {
    // Not empty
}
const parseFileUnit = (size) => {
    if (typeof size === 'string')
        switch (size.slice(-1)) {
            case 'k':
                return +size.slice(0, size.length - 1) * 1024;
            case 'm':
                return +size.slice(0, size.length - 1) * 1048576;
            default:
                return +size;
        }
    return size;
};
const validateFile = (options, value) => {
    if (!(value instanceof Blob))
        return false;
    if (options.minSize && value.size < parseFileUnit(options.minSize))
        return false;
    if (options.maxSize && value.size > parseFileUnit(options.maxSize))
        return false;
    if (options.extension)
        if (typeof options.extension === 'string') {
            if (!value.type.startsWith(options.extension))
                return false;
        }
        else {
            for (let i = 0; i < options.extension.length; i++)
                if (value.type.startsWith(options.extension[i]))
                    return true;
            return false;
        }
    return true;
};
const Files = system_1.TypeSystem.Type('Files', (options, value) => {
    if (!Array.isArray(value))
        return validateFile(options, value);
    if (options.minItems && value.length < options.minItems)
        return false;
    if (options.maxItems && value.length > options.maxItems)
        return false;
    for (let i = 0; i < value.length; i++)
        if (!validateFile(options, value[i]))
            return false;
    return true;
});
typebox_1.FormatRegistry.Set('numeric', (value) => !!value && !isNaN(+value));
typebox_1.FormatRegistry.Set('ObjectString', (value) => {
    let start = value.charCodeAt(0);
    // If starts with ' ', '\t', '\n', then trim first
    if (start === 9 || start === 10 || start === 32)
        start = value.trimStart().charCodeAt(0);
    if (start !== 123 && start !== 91)
        return false;
    try {
        JSON.parse(value);
        return true;
    }
    catch {
        return false;
    }
});
exports.ElysiaType = {
    Numeric: (property) => {
        const schema = typebox_1.Type.Number(property);
        return t
            .Transform(t.Union([
            t.String({
                format: 'numeric',
                default: 0
            }),
            t.Number(property)
        ], property))
            .Decode((value) => {
            const number = +value;
            if (isNaN(number))
                return value;
            if (property && !value_1.Value.Check(schema, number))
                throw new error_1.ValidationError('property', schema, number);
            return number;
        })
            .Encode((value) => value);
    },
    ObjectString: (properties, options) => t
        .Transform(t.Union([
        t.String({
            format: 'ObjectString',
            default: ''
        }),
        t.Object(properties, options)
    ], options))
        .Decode((value) => {
        if (typeof value === 'string')
            try {
                return JSON.parse(value);
            }
            catch {
                return value;
            }
        return value;
    })
        .Encode((value) => JSON.stringify(value)),
    File: system_1.TypeSystem.Type('File', validateFile),
    Files: (options = {}) => t
        .Transform(t.Union([Files(options)]))
        .Decode((value) => {
        if (Array.isArray(value))
            return value;
        return [value];
    })
        .Encode((value) => value),
    Nullable: (schema) => t.Union([t.Null(), schema]),
    /**
     * Allow Optional, Nullable and Undefined
     */
    MaybeEmpty: (schema) => t.Union([t.Null(), t.Undefined(), schema]),
    Cookie: (properties, options) => t.Object(properties, options)
};
t.ObjectString = exports.ElysiaType.ObjectString;
/**
 * A Numeric string
 *
 * Will be parse to Number
 */
t.Numeric = exports.ElysiaType.Numeric;
t.File = (arg = {}) => exports.ElysiaType.File({
    default: 'File',
    ...arg,
    extension: arg?.type,
    type: 'string',
    format: 'binary'
});
t.Files = (arg = {}) => exports.ElysiaType.Files({
    ...arg,
    elysiaMeta: 'Files',
    default: 'Files',
    extension: arg?.type,
    type: 'array',
    items: {
        ...arg,
        default: 'Files',
        type: 'string',
        format: 'binary'
    }
});
t.Nullable = (schema) => exports.ElysiaType.Nullable(schema);
t.MaybeEmpty = exports.ElysiaType.MaybeEmpty;
t.Cookie = exports.ElysiaType.Cookie;
__exportStar(require("@sinclair/typebox/system"), exports);
__exportStar(require("@sinclair/typebox/compiler"), exports);
// type Template =
// 	| string
// 	| number
// 	| bigint
// 	| boolean
// 	| StringConstructor
// 	| NumberConstructor
// 	| undefined
// type Join<A> = A extends Readonly<[infer First, ...infer Rest]>
// 	? (
// 			First extends Readonly<Template[]>
// 				? First[number]
// 				: First extends StringConstructor
// 				? string
// 				: First extends NumberConstructor
// 				? `${number}`
// 				: First
// 	  ) extends infer A
// 		? Rest extends []
// 			? A extends undefined
// 				? NonNullable<A> | ''
// 				: A
// 			: // @ts-ignore
// 			A extends undefined
// 			? `${NonNullable<A>}${Join<Rest>}` | ''
// 			: // @ts-ignore
// 			  `${A}${Join<Rest>}`
// 		: ''
// 	: ''
// const template = <
// 	const T extends Readonly<(Template | Readonly<Template[]>)[]>
// >(
// 	...p: T
// ): Join<T> => {
// 	return a as any
// }
// const create =
// 	<const T extends string>(t: T): ((t: T) => void) =>
// 	(t) =>
// 		t
// const optional = <
// 	const T extends Readonly<(Template | Readonly<Template[]>)[]>
// >(
// 	...p: T
// ): T | undefined => {
// 	return undefined
// }
// template.optional = optional
// const hi = create(
// 	template(
// 		['seminar', 'millennium'],
// 		':',
// 		['Rio', 'Yuuka', 'Noa', 'Koyuki'],
// 		template.optional(template(',', ['Rio', 'Yuuka', 'Noa', 'Koyuki'])),
// 		template.optional(template(',', ['Rio', 'Yuuka', 'Noa', 'Koyuki'])),
// 		template.optional(template(',', ['Rio', 'Yuuka', 'Noa', 'Koyuki']))
// 	)
// )
// hi(`seminar:Noa,Koyuki,Yuuka`)
// const a = TypeCompiler.Compile(t.String())
// console.log(v.Decode.toString())
// const T = t.Transform(v.schema)
// 	.Decode((value) => new Date(value)) // required: number to Date
// 	.Encode((value) => value.getTime()) // required: Date to number
// const decoded = Value.Decode(T, 0) // const decoded = Date(1970-01-01T00:00:00.000Z)
// const encoded = Value.Encode(T, decoded)
